src/backends/r1cs/mod.rs

```diff
@@ -312,7 +312,12 @@ where
                     // Defer calculation for output vars.
                     // The reasoning behind this is to avoid deep recursion potentially triggered by the public output var at the beginning.
                     Value::PublicOutput(_) => Ok(F::zero()),
-                    _ => self.compute_val(witness_env, val, index),
+                    _ => {
+                        // wrap as CellVar
+                        // todo: we might have a cache for CellVar somewhere to retrieve the span by var index
+                        let var = R1csCellVar::new(index, Span::default());
+                        self.compute_var(witness_env, var)
```

Comment:
this feels very hacky to me

Reply:
my intuition is that the cache is not going to work properly because of the vars not necessarily matching (because of a different span, for example)

I think we need to use a usize for the cache, as such I think we need to move `compute_var` to the backend, does that make sense?

Reply:
The problem is the `compute_var` contains the shared code to calculate the values across different backends. Moving the `compute_var` might mean making this part duplicated in different backends.

Reply:
We might still need to keep the compute_val functions, and have the backends to implement the compute_var. 

So this allows the backends to decide how to load the Value before passing it to compute_val. This hides the details of CellVar such as index, and cache could use the index as the key.

Reply:
[a6ed103](https://github.com/zksecurity/noname/pull/69/commits/a6ed10355026028da43d73551935ccf08368c267)